src/DS18XXX/shelly_temp_sensor_ow.cpp

```diff
@@ -0,0 +1,227 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "shelly_temp_sensor_ow.hpp"
+
+#include <cmath>
+
+#include "mgos_gpio.h"
+#include "mgos_onewire.h"
+
+/* Model IDs */
+#define DS18S20MODEL 0x10
+#define DS18B20MODEL 0x28
+#define DS1822MODEL 0x22
+#define DS1825MODEL 0x3B
+#define DS28EA00MODEL 0x42
+
+/* Device resolution */
+#define TEMP_9_BIT 0x1F  /* 9 bit */
+#define TEMP_10_BIT 0x3F /* 10 bit */
+#define TEMP_11_BIT 0x5F /* 11 bit */
+#define TEMP_12_BIT 0x7F /* 12 bit */
+
+#define CONVERT_T 0x44
+#define READ_SCRATCHPAD 0xBE
+#define READ_POWER_SUPPLY 0xB4
+
+struct __attribute__((__packed__)) Scratchpad {
+  int16_t temperature;
+  uint8_t th;
+  uint8_t tl;
+  uint8_t configuration;
+  uint8_t rfu;
+  uint8_t count_remain;
+  uint8_t count_per_c;
+  uint8_t crc;
+};
+
+struct __attribute__((__packed__)) ROM {
+  uint8_t family;
+  uint64_t serial;
+  uint8_t crc;
+};
+
+namespace shelly {
+
+Onewire::Onewire(int pin_in, int pin_out) {
+  pin_out_ = pin_out;
+  ow_ = mgos_onewire_create_separate_io(pin_in, pin_out);
+}
+
+Onewire::~Onewire() {
+  mgos_onewire_close(ow_);
+
+  // release output pin
+  mgos_gpio_setup_input(pin_out_, MGOS_GPIO_PULL_UP);
+}
+
+struct mgos_onewire *Onewire::Get() {
+  return ow_;
+}
+
+std::vector<std::unique_ptr<TempSensor>> Onewire::DiscoverAll() {
+  std::vector<std::unique_ptr<TempSensor>> sensors;
+  mgos_onewire_search_clean(ow_);
+  std::unique_ptr<TempSensor> sensor;
+  while (sensor = NextAvailableSensor(0)) {
+    sensors.push_back(std::move(sensor));
+  }
+  LOG(LL_INFO, ("Discovered %i sensors", sensors.size()));
+  return sensors;
+}
+
+std::unique_ptr<TempSensor> Onewire::NextAvailableSensor(int type) {
+  struct ROM rom = {0};
+  int mode = 0;
+  std::unique_ptr<TempSensor> sensor;
+  if (mgos_onewire_next(ow_, (uint8_t *) &rom, mode)) {
+    uint8_t family = rom.family;
+    if (TempSensorDS18XXX::SupportsFamily(family)) {
+      sensor.reset(new TempSensorDS18XXX(ow_, &rom));
+    } else {
+      LOG(LL_INFO, ("Found non-supported device"));
+    }
+  }
+  (void) type;
+  return sensor;
+}
+
+void TempSensorDS18XXX::StartUpdating(int interval) {
+  read_timer_.Clear();  // Clear eventually pending read
+  meas_timer_.Reset(interval, MGOS_TIMER_REPEAT | MGOS_TIMER_RUN_NOW);
+}
+
+TempSensorDS18XXX::TempSensorDS18XXX(struct mgos_onewire *ow,
+                                     struct ROM *rom_content)
+    : cached_temperature_(0),
+      meas_timer_(std::bind(&TempSensorDS18XXX::UpdateTemperatureCB, this)),
+      read_timer_(std::bind(&TempSensorDS18XXX::ReadTemperatureCB, this)) {
+  rom_ = new ROM();
```

Comment:
`rom_` doesn't have to be a pointer, you can make it a straight `struct ROM` and use default copy constructor to copy, no memcpy needed, and no explicit delete at the end.